File: Lab2/FileReverse.java
//---------------------------------------------------
// Name: Bingxue Ouyang
// CruzID: bouyang
// FileReverse.java
// Reverse input file text
//---------------------------------------------------
import java.io.*;
import java.util.Scanner;

class FileReverse{
    public static void main(String[] args) throws IOException{
        // check that command contains input file and output file
        if(args.length<2){
            System.out.println("Usage: java -jar FileReverse.jar <input file> <output file>");
            System.exit(1);
        }


        // open input and output files
        Scanner in = new Scanner(new File(args[0]));
        PrintWriter out = new PrintWriter(new FileWriter(args[1]));
        String result="";

        // read input file
        while(in.hasNextLine()){
        // trim spaces and add trailing space for split()
            String line = in.nextLine().trim()+" ";
			
            // split line around white space
            String[] split = line.split("\\s+");
			
			// call reverse function and print out
			for(int j=0; j<split.length; j++){
				out.println(stringReverse(split[j]));
			}		
		}
		
		// close files
		in.close();
		out.close();	
	}
		
	public static String stringReverse(String s){
		String reverse = "";
		
		// split characters and reverse
		for(int i=s.length(); i>0; i--){
			reverse += s.charAt(i-1);
		}
		return reverse;
	}
}

File: Lab3/fileReverse.c
/* 
 * Name: Bingxue Ouyang
 * cruzID: bouyang
 * fileReverse.c
 * a c program that reads input file and reverse it to output file
 */

#include<stdio.h>
#include<stdlib.h>
#include<string.h>

void stringReverse(char* s){
    // the first char
    int head = 0;
    // the last char
    int end = strlen(s)-1;
    // start replacing from both sides, move towards middle
    while(head<=end){
        char replace = s[end];
        s[end] = s[head];
        s[head] = replace;
        head++;
        end--;
    }
}
int main(int argc, char* argv[]){
    // check if command contains input and output file
    if(argc<2){
        printf("Usage: %s some-string\n", argv[0]);
        exit(EXIT_FAILURE);
    }
    // initialize input and output files
    FILE* in = fopen(argv[1], "r");
    FILE* out = fopen(argv[2], "w");
    char word[256];
    // check if input and output files are able to read and write
    if(in==NULL){
        printf("Unable to read from file %s\n", argv[1]);
        exit(EXIT_FAILURE);
    }
    if(out==NULL){
        printf("Unable to write to file %s\n", argv[2]);
        exit(EXIT_FAILURE);
    }
    // start reading and reverse
    while(fscanf(in, " %s", word) != EOF){
        stringReverse(word);
        fprintf(out,"%s\n", word);
    }
    // close both files
    fclose(in);
    fclose(out);
    
    return (EXIT_SUCCESS);
}